lib/promutils/labelscompressor.go

```diff
@@ -59,8 +59,16 @@ func (lc *LabelsCompressor) compress(dst []uint64, labels []prompbmarshal.Label)
 			idx := lc.nextIdx.Add(1)
 			v = idx
 			labelCopy := cloneLabel(label)
+
+			// must store idxToLabel entry before labelToIdx,
+			// so it can be found by possible concurrent goroutines.
 			lc.idxToLabel.Store(idx, labelCopy)
-			lc.labelToIdx.Store(labelCopy, v)
+			vNew, loaded := lc.labelToIdx.LoadOrStore(labelCopy, v)
```

Comment:
what do you think about using LoadOrStore(label, -1) at line 57 and after `if !ok` add `else if v == -1 {goto a place before line 57}`?

Reply:
@hagen1778 Yes, it will, but that should be fine.
If both index 99 and 100 are assigned to label `foo=1` in two concurrent pushes, we should always return the oldest index 99 to ensure both samples are compressed to the same key.
Since we need to store idxToLabel first to ensure the idx can be searched after `lc.labelToIdx.LoadOrStore`, `lc.idxToLabel` could contain a duplicated entries `[100]="foo=1"`. We can also set the entry to nil in `lc.idxToLabel` as `[100]=nil` after the check, but given the low likelihood of this issue, I didn't implement it. I can add it if you think it's better to understand.

>what do you think about using LoadOrStore(label, -1) at line 57 and after if !ok add else if v == -1 {goto a place before line 57}?

That won't fix the issue. The issue is we need to store idxToLabel before labelToIdx, so after storing idxToLabel, the labelToIdx entry could be created by other concurrent goroutine. And in that case, we must re-use it.

Reply:
I see. Looks reasonable to me. I'd prefer comment to be specific, that it would result into different encoded keys and will affect aggregation logic.

Reply:
@AndrewChubatiuk do you see any other issues with this fix? Does it sound reasonable to you?